### PR TITLE
[SYCL] Add support for radix_sorter

### PR DESCRIPTION
**Pull request by romanovvlad**
Co-authored-by: Andrei Fedorov <andrey.fedorov@intel.com>

### CODE REVIEW

**Comment by romanovvlad** [comment]
Test is here: https://github.com/intel/llvm-test-suite/pull/1435

**Comment by romanovvlad** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1435

**Comment by romanovvlad** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1435

**Comment by romanovvlad** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1435/

**Comment by romanovvlad** [comment]
@v-klochkov Could you please take a look?

**Review by v-klochkov — DISMISSED** [review]
I have only few minor non-blocking comments.

File: sycl/include/sycl/detail/group_sort_impl.hpp [hunk 1]
```diff
@@ -250,6 +251,412 @@ void merge_sort(Group group, Iter first, const std::size_t n, Compare comp,
...
+                                      std::is_unsigned<ValT>::value>> {
+  using Type = ValT;
+};
+
+// for signed integrals or floatings we map: size -> corresponding unsigned
+// integral
+template <typename ValT>
+struct Ordered<ValT, std::enable_if_t<(std::is_integral<ValT>::value &&
+                                       std::is_signed<ValT>::value) ||
+                                      !std::is_integral<ValT>::value>> {
```
[the remark this answers is not in the document]
Comment (romanovvlad) [reply sycl/include/sycl/detail/group_sort_impl.hpp:332]: Done.

File: sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp [hunk 2]
```diff
@@ -92,6 +93,100 @@ template <typename Compare = std::less<>> class default_sorter {
...
+
+public:
+  template <std::size_t Extent>
+  radix_sorter(sycl::span<std::byte, Extent> scratch_,
+               const std::bitset<sizeof(ValT) *CHAR_BIT> mask =
+                   std::bitset<sizeof(ValT) * CHAR_BIT>(
+                       std::numeric_limits<unsigned long long>::max()))
+      : scratch(scratch_.data()), scratch_size(scratch_.size()) {
+    static_assert((std::is_arithmetic<ValT>::value ||
+                   std::is_same<ValT, sycl::half>::value),
```
[the remark this answers is not in the document]
Comment (romanovvlad) [reply sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp:129]: 1. C++ reference says it's UB to specialize `std::is_arithmetic` for custom types: https://en.cppreference.com/w/cpp/types/is_arithmetic

2. No, fixed.

File: sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp [hunk 3]
```diff
@@ -92,6 +93,100 @@ template <typename Compare = std::less<>> class default_sorter {
...
+    ValT result[]{val};
+    sycl::detail::privateMemorySort</*is_key_value=*/false,
+                                    /*is_blocked=*/true,
+                                    /*items_per_work_item=*/1, bits>(
+        g, result, /*empty*/ result,
+        typename detail::ConvertToComp<ValT, OrderT>::Type{}, scratch,
+        first_bit, last_bit);
+    return result[0];
+#else
+    return ValT{};
```
[the remark this answers is not in the document]
Comment (romanovvlad) [reply sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp:167]: Add throw of an exception.

File: sycl/include/sycl/detail/group_sort_impl.hpp [hunk 4]
```diff
@@ -361,7 +365,9 @@ convertToOrdered(ValT value) {
 // converts floating type to Ordered (in terms of bitness) type
 template <typename ValT>
 std::enable_if_t<!std::is_same<ValT, OrderedT<ValT>>::value &&
-                     !std::is_integral<ValT>::value,
+                         std::is_floating_point<ValT>::value ||
+                     std::is_same<ValT, sycl::half>::value ||
+                     std::is_same<ValT, sycl::ext::oneapi::bfloat16>::value,
```
Comment (v-klochkov) [inline sycl/include/sycl/detail/group_sort_impl.hpp:370]: Shouldn't it be in ( ) ? I.e.:
```suggestion
std::enable_if_t<!std::is_same_v<ValT, OrderedT<ValT>>::value &&
                         (std::is_floating_point_v<ValT> ||
                     std::is_same_v<ValT, sycl::half> ||
                     std::is_same_v<ValT, sycl::ext::oneapi::bfloat16>)
```

Comment (romanovvlad) [reply sycl/include/sycl/detail/group_sort_impl.hpp]: Fixed. Thank you.

**Comment by romanovvlad** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1435

**Review by v-klochkov — APPROVED** [review]

File: sycl/include/sycl/detail/group_sort_impl.hpp [hunk 5]
```diff
@@ -365,9 +365,9 @@ convertToOrdered(ValT value) {
 // converts floating type to Ordered (in terms of bitness) type
 template <typename ValT>
 std::enable_if_t<!std::is_same<ValT, OrderedT<ValT>>::value &&
-                         std::is_floating_point<ValT>::value ||
-                     std::is_same<ValT, sycl::half>::value ||
-                     std::is_same<ValT, sycl::ext::oneapi::bfloat16>::value,
+                     (std::is_floating_point<ValT>::value ||
+                      std::is_same<ValT, sycl::half>::value ||
+                      std::is_same<ValT, sycl::ext::oneapi::bfloat16>::value),
```
Comment (aelovikov-intel) [inline sycl/include/sycl/detail/group_sort_impl.hpp:370]: Don't want to block this, but we have `std::is_same_v` from C++17 unconditionally now.

Comment (romanovvlad) [reply sycl/include/sycl/detail/group_sort_impl.hpp]: Done.

**Comment by romanovvlad** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1435

**Comment by romanovvlad** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1435

**Comment by romanovvlad** [comment]
The failures are unrelated.

File: sycl/include/sycl/detail/group_sort_impl.hpp [hunk 6]
```diff
@@ -250,6 +252,416 @@ void merge_sort(Group group, Iter first, const std::size_t n, Compare comp,
...
+  return std::numeric_limits<ValT>::max();
+}
+
+template <typename ValT> ValT getDefaultValue(std::greater<ValT>) {
+  return std::numeric_limits<ValT>::lowest();
+}
+
+template <bool is_key_value_sort> struct ValuesAssigner {
+  template <typename IterInT, typename IterOutT>
+  void operator()(IterOutT output, size_t idx_out, IterInT input,
```
[the remark this answers is not in the document]
Comment (romanovvlad) [reply sycl/include/sycl/detail/group_sort_impl.hpp:426]: Could you please clarify why?

File: sycl/include/sycl/detail/group_sort_impl.hpp [hunk 7]
```diff
@@ -250,6 +252,416 @@ void merge_sort(Group group, Iter first, const std::size_t n, Compare comp,
   }
 }
 
+// traits for ascending functors
+template <typename CompT> struct IsCompAscending {
+  static constexpr bool value = false;
+};
+template <typename Type> struct IsCompAscending<std::less<Type>> {
```
[the remark this answers is not in the document]
Comment (romanovvlad) [reply sycl/include/sycl/detail/group_sort_impl.hpp:259]: No, we do not support `sycl::less` anymore.

File: sycl/include/sycl/detail/group_sort_impl.hpp [hunk 8]
```diff
@@ -250,6 +252,416 @@ void merge_sort(Group group, Iter first, const std::size_t n, Compare comp,
...
+  uint32_t *scan_memory = reinterpret_cast<uint32_t *>(memory);
+  for (uint32_t state = 0; state < radix_states; ++state)
+    scan_memory[state * wgsize + idx] = 0;
+
+  sycl::group_barrier(group);
+
+  // 1.2. count values and write result to private count array and count memory
+
+  for (uint32_t i = 0; i < items_per_work_item; ++i) {
+    const uint32_t val_idx = items_per_work_item * idx + i;
```
[the remark this answers is not in the document]
Comment (romanovvlad) [reply sycl/include/sycl/detail/group_sort_impl.hpp:472]: It should be optimized by the compiler.

File: sycl/include/sycl/detail/group_sort_impl.hpp [hunk 9]
```diff
@@ -250,6 +252,416 @@ void merge_sort(Group group, Iter first, const std::size_t n, Compare comp,
...
+    reduced += scan_memory[idx * radix_states + i];
+
+  // 2.2. Exclusive scan: over work items
+  uint32_t scanned =
+      sycl::exclusive_scan_over_group(group, reduced, std::plus<uint32_t>());
+
+  // 2.3. Exclusive downsweep: exclusive scan over radix states
+  for (uint32_t i = 0; i < radix_states; ++i) {
+    uint32_t value = scan_memory[idx * radix_states + i];
+    scan_memory[idx * radix_states + i] = scanned;
```
[the remark this answers is not in the document]
Comment (romanovvlad) [reply sycl/include/sycl/detail/group_sort_impl.hpp:499]: It should be optimized by the compiler.

File: sycl/include/sycl/detail/group_sort_impl.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (romanovvlad) [reply sycl/include/sycl/detail/group_sort_impl.hpp:426]: Removed `std::` to align the code.

File: sycl/include/sycl/detail/group_sort_impl.hpp [hunk 10]
```diff
@@ -250,6 +252,416 @@ void merge_sort(Group group, Iter first, const std::size_t n, Compare comp,
...
+template <> struct ValuesAssigner<false> {
+  template <typename IterInT, typename IterOutT>
+  void operator()(IterOutT, size_t, IterInT, size_t) {}
+
+  template <typename IterOutT, typename ValT>
+  void operator()(IterOutT, size_t, ValT) {}
+};
+
+// The iteration of radix sort for unknown number of elements per work item
+template <uint32_t radix_bits, bool is_key_value_sort, typename KeysT,
```
[the remark this answers is not in the document]
Comment (romanovvlad) [reply sycl/include/sycl/detail/group_sort_impl.hpp:446]: Removed `std::` to align the code.

File: sycl/include/sycl/detail/group_sort_impl.hpp [hunk 11]
```diff
@@ -250,6 +252,416 @@ void merge_sort(Group group, Iter first, const std::size_t n, Compare comp,
...
+    auto val = convertToOrdered((val_idx < n) ? keys_input[val_idx]
+                                              : getDefaultValue(comp));
+    // get bit values in a certain bucket of a value
+    uint32_t bucket_val =
+        getBucketValue<radix_bits, is_comp_asc>(val, radix_iter);
+
+    uint32_t new_offset_idx = private_scan_memory[bucket_val]++ +
+                              scan_memory[bucket_val * wgsize + idx];
+    if (val_idx < n) {
+      keys_output[new_offset_idx] = keys_input[val_idx];
```
[the remark this answers is not in the document]
Comment (andreyfe1) [reply sycl/include/sycl/detail/group_sort_impl.hpp:514]: Sorry, we use just arithmetic types for radix. It's ok if we work with trivially copyable types

**Review by andreyfe1 — APPROVED** [review]

**Comment by romanovvlad** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1435

**Comment by pvchupin** [comment]
@romanovvlad, post-commit issue on windows https://github.com/intel/llvm/actions/runs/3687647938/jobs/6241495307

**Comment by againull** [comment]
> @romanovvlad, post-commit issue on windows https://github.com/intel/llvm/actions/runs/3687647938/jobs/6241495307

This seemed familiar, so I prepared a PR: https://github.com/intel/llvm/pull/7768

**Pull request merged**